src/core_modules/capture-core/metadataRetrieval/orgUnitName/orgUnitName.js

```diff
@@ -130,44 +143,71 @@ export const useOrgUnitNames = (orgUnitIds: Array<string>): {
 };
 
 export async function getOrgUnitNames(orgUnitIds: Array<string>, querySingleResource: QuerySingleResource): Promise<{|
-    [orgUnitId: string]: {|
-        id: string,
+[orgUnitId: string]: {|
+    id: string,
         displayName: string,
     |}
 |}> {
     await Promise.all(createBatches(orgUnitIds)
         .map(batch => querySingleResource(displayNamesQuery.organisationUnits, { filter: batch.join(',') })
             .then(({ organisationUnits }) => {
-                for (const { id, displayName } of organisationUnits) {
-                    displayNameCache[id] = displayName;
-                }
-            })));
+                updateCacheWithOrgUnits(organisationUnits);
+            }),
+        ),
+    );
 
     return orgUnitIds.reduce((acc, orgUnitId) => {
         acc[orgUnitId] = {
             id: orgUnitId,
-            name: displayNameCache[orgUnitId],
+            name: orgUnitCache[orgUnitId]?.displayName,
         };
         return acc;
     }, {});
 }
 
-export const useOrgUnitName = (orgUnitId: ?string): {
+export const useOrgUnitNameWithAncestors = (orgUnitId: ?string): {
     displayName?: string,
-    error?: any,
+    ancestors?: Array<{| id: string, displayName: string |}>,
+    error ?: any,
 } => {
-    const cachedOrgUnitName = orgUnitId && displayNameCache[orgUnitId];
-    const fetchId = cachedOrgUnitName ? undefined : orgUnitId;
-    const { orgUnit, error } = useOrganisationUnit(fetchId, 'displayName');
-    if (cachedOrgUnitName) {
-        return { displayName: cachedOrgUnitName };
-    } else if (orgUnit && fetchId) {
-        displayNameCache[orgUnit.id] = orgUnit.displayName;
-        if (orgUnit.id === fetchId) {
-            return { displayName: orgUnit.displayName, error };
-        }
+    const cachedOrgUnit = orgUnitId && orgUnitCache[orgUnitId];
+    const fetchId = cachedOrgUnit ? undefined : orgUnitId;
+    const { orgUnit: fetchedOrgUnit, error } = useOrganisationUnit(fetchId, 'displayName,ancestors[id,displayName]');
+
+    if (orgUnitId && cachedOrgUnit) {
+        const getAncestors = (parentOrgUnitId) => {
+            const orgUnit = orgUnitCache[parentOrgUnitId];
+
+            if (!orgUnit) return [];
+
+            const ancestors = getAncestors(orgUnit.ancestor);
+            ancestors.push({
+                displayName: orgUnit.displayName,
+                id: orgUnitId,
+            });
+
+            return ancestors;
+        };
+
+        const ancestors = getAncestors(orgUnitId);
+        ancestors.pop();
```

Comment:
NIT: I would prefer not to pop for better readability. I think we can do this:
```suggestion
        const ancestors = cachedOrgUnit.ancestor ? getAncestors(cachedOrgUnit.ancestor) : [];
```

Reply:
I would also suggest moving `getAncestors` out into global scope. In this case I think it will reveal an error (`orgUnitId` not defined).